osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatisticTooltip.cs

```diff
@@ -0,0 +1,158 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Extensions.LocalisationExtensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Cursor;
+using osu.Framework.Graphics.Shapes;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Rulesets.Difficulty;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Game.Screens.Ranking.Expanded.Statistics
+{
+    using static PerformanceStatistic;
+
+    public class PerformanceStatisticTooltip : VisibilityContainer, ITooltip<PerformanceBreakdown>
+    {
+        private readonly Box background;
+        private Colour4 titleColor;
+        private Colour4 textColour;
+
+        protected override Container<Drawable> Content { get; }
+
+        public PerformanceStatisticTooltip()
+        {
+            AutoSizeAxes = Axes.Both;
+            Masking = true;
+            CornerRadius = 5;
+
+            InternalChildren = new Drawable[]
+            {
+                background = new Box
+                {
+                    RelativeSizeAxes = Axes.Both
+                },
+                Content = new FillFlowContainer
+                {
+                    AutoSizeAxes = Axes.Both,
+                    Direction = FillDirection.Vertical,
+                    Padding = new MarginPadding { Left = 10, Right = 10, Top = 5, Bottom = 5 }
+                },
+            };
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours)
+        {
+            background.Colour = colours.Gray3;
+            titleColor = colours.Blue;
+            textColour = colours.BlueLighter;
+        }
+
+        protected override void PopIn()
+        {
+            // Don't display the tooltip if "Total" is the only item
+            if (currentPerformance.Performance.GetAttributesForDisplay().Count() > 1)
+                this.FadeIn(200, Easing.OutQuint);
+        }
+
+        protected override void PopOut() => this.FadeOut(200, Easing.OutQuint);
+
+        private PerformanceBreakdown currentPerformance;
+
+        public void SetContent(PerformanceBreakdown performance)
+        {
+            if (performance == currentPerformance)
+                return;
+
+            currentPerformance = performance;
+
+            UpdateDisplay(performance);
+        }
+
+        private Drawable createAttributeItem(PerformanceDisplayAttribute attribute, PerformanceDisplayAttribute perfectAttribute)
+        {
+            bool isTotal = attribute.PropertyName == nameof(PerformanceAttributes.Total);
+            float fraction = (float)(attribute.Value / perfectAttribute.Value);
+            if (float.IsNaN(fraction))
+                fraction = 0;
```

Comment:
this is better, but isn't quite what i had in mind [in my earlier comment](https://github.com/ppy/osu/pull/16483#issuecomment-1014627537). while displaying the user's score performance attributes as a ratio relative to the perfect score's, there's still sort of an issue in that there is no actual guarantee that the pp algorithms are going to give equal weighting of the attributes, or that the final pp value is going to be a linear combination of the attribute values.

to substantiate this: here's how 3/4 "official" rulesets compute the final pp currently.

https://github.com/ppy/osu/blob/b3983db69d5a5646cae2c3bdd9db326ea472fbfd/osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs#L66-L72
https://github.com/ppy/osu/blob/b3983db69d5a5646cae2c3bdd9db326ea472fbfd/osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs#L48-L52
https://github.com/ppy/osu/blob/b3983db69d5a5646cae2c3bdd9db326ea472fbfd/osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs#L66-L70

notice that these all use the [generalized / power mean](https://en.wikipedia.org/wiki/Generalized_mean) of the attribute values to compute the final value. and the thing with the power mean is that it's non linear.

below is a graph; red illustrates the power mean, and blue illustrates the linear sum. the interpretation of this graph is: "each point on the red/blue curve is such that if the x and y coordinates are plugged into the correct curve's formula, the result is 10". in other words, moving that to the pp domain - that's what the values of performance attributes would have to be for the final pp value to be 10 according to each curve's formula.

![2022-01-18-223159_965x430_scrot](https://user-images.githubusercontent.com/20418176/150022304-b9a417a3-1969-490a-815f-292e52f958da.png)

note that the red curve is not linear flat, it's slightly convex. that's the nonlinearity.

to get to the point: that's why i proposed the original scheme, to eliminate the impact of the choice of combining function on how the breakdown works. because imagine that you may want to raise an attribute to another attribute's power, or similar - that isn't going to translate well to the current implementation, because seemingly small variations of a performance attribute may wildly influence the final pp value. so what i propose instead is, formally:

let f(x<sub>1</sub>, x<sub>2</sub>, ..., x<sub>n</sub>) be the function used to combine values of performance attributes x<sub>1</sub>, x<sub>2</sub>, ..., x<sub>n</sub> into a final pp value; in our current situation, it would be the power mean. now let p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>n</sub> be the value of performance attributes for a perfect play on the same map. i propose that the breakdown's value for the i-th attribute should equal:

b<sub>i</sub> = f(p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>i - 1</sub>, x<sub>i</sub>, p<sub>i + 1</sub> ..., p<sub>n</sub>) / f(p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>i - 1</sub>, p<sub>i</sub>, p<sub>i + 1</sub> ..., p<sub>n</sub>)

this would *roughly* represent the number of pp that the player lost on the i-th attribute alone if they played perfectly with respect to every other attribute, compared to if they were playing 100% perfectly in general.

i think this is the best that can be done with respect to separating out attributes' influence on the final pp value, without constraining how the combining function f works and looks.

Reply:
had some discussion about this elsewhere and it was rightly pointed out to me that the above can have a weird artifact wherein fixing the i-th attribute to 0 might not yield 0 with the proposed scheme. this could be remedied by having that as the absolute minimum (as in, the evaluation scale goes from "perfect score with i-th attribute fixed to 0" to "perfect score"), i.e. in the formula above, you'd subtract f(p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>i - 1</sub>, 0, p<sub>i + 1</sub> ..., p<sub>n</sub>) from both the numerator and the denominator.

Reply:
I do agree that this is a better scheme than simple ratios, but the performance calculator needs to be refactored for this scheme to work. I'd rather do this in a separate PR after this one's merged.